When suggesting to import an item, also suggest changing the path if appropriate

estebank opened the pull request:
When we don't find an item we search all of them for an appropriate
import and suggest `use`ing it. This is sometimes done for expressions
that have paths with more than one segment. We now also suggest changing
that path to work with the `use`.

Fix #95413

estebank committed:
When suggesting to import an item, also suggest changing the path if appropriate

When we don't find an item we search all of them for an appropriate
import and suggest `use`ing it. This is sometimes done for expressions
that have paths with more than one segment. We now also suggest changing
that path to work with the `use`.

Fix #95413

rust-highfive commented:
r? @compiler-errors

(rust-highfive has picked a reviewer for you, use r? to override)

est31 commented:
I think the ... notation is not easy to parse, given all the other punctuation that's present. What about "Consider importing these items instead / You can then refer to them directly"?

compiler-errors reviewed (approved):
Not sure if I have a strong opinion on the "... then" notation as opposed to more explicit phrasing, since we have examples of other suggestions split up by `...`. 

Feel free to refine it if you want, or r=me if not.

est31 commented:
@compiler-errors your comment has made me look for existing examples in the compiler. There were a bunch.

```
error: multiple patterns overlap on their endpoints
  --> $DIR/overlapping_range_endpoints.rs:16:22
   |
LL |     m!(0u8, 30..=40, 20..=30);
   |             -------  ^^^^^^^ ... with this range
   |             |
   |             this range overlaps on `30_u8`...
   |
   = note: you likely meant to write mutually exclusive ranges
```

```
error[E0746]: return type cannot have an unboxed trait object
  --> $DIR/point-to-type-err-cause-on-impl-trait-return.rs:77:13
   |
LL | fn pug() -> dyn std::fmt::Display {
   |             ^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = note: for information on trait objects, see <https://doc.rust-lang.org/book/ch17-02-trait-objects.html#using-trait-objects-that-allow-for-values-of-different-types>
   = note: if all the returned values were of the same type you could use `impl std::fmt::Display` as the return type
   = note: for information on `impl Trait`, see <https://doc.rust-lang.org/book/ch10-02-traits.html#returning-types-that-implement-traits>
   = note: you can create a new `enum` with a variant for each returned type
help: return a boxed trait object instead
   |
LL | fn pug() -> Box<dyn std::fmt::Display> {
   |             ++++                     +
help: ... and box this value
   |
LL |         0 => Box::new(0i32),
   |              +++++++++    +
help: ... and box this value
   |
LL |         1 => Box::new(1u32),
   |              +++++++++    +
help: ... and box this value
   |
LL |         _ => Box::new(2u32),
   |              +++++++++    +
```

```
error[E0499]: cannot borrow `*heap` as mutable more than once at a time
  --> $DIR/issue-85581.rs:9:22
   |
LL |     match heap.peek_mut() {
   |           ---------------
   |           |
   |           first mutable borrow occurs here
   |           a temporary with access to the first borrow is created here ...
LL |         Some(_) => { heap.pop(); },
   |                      ^^^^^^^^^^ second mutable borrow occurs here
...
LL | }
   | - ... and the first borrow might be used here, when that temporary is dropped and runs the destructor for type `Option<PeekMut<'_, i32>>`
```

They read more clearly to me than 

```
help: consider importing this function...
   |
   |
LL | use hi_str;
LL | use hi_str;
   |
   |
help: ...and refer to it directly
   |
LL -     println!("{}", circular_modules_main::hi_str());
LL +     println!("{}", hi_str());
   | 
```

Maybe it's because of the suggested diff that's being printed, without the ^^^^ underlining?

est31 commented:
(Note that I don't feel strongly about it, this PR is an improvement even in the current state)

estebank commented:
I'm ok with changing the wording

bors commented:
:umbrella: The latest upstream changes (presumably #90988) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
Tweak wording

estebank commented:
Rebased and changed the wording to avoid the `...`s.

Re the rendered output being weird, is because of the specialized removal "diff" output. It looks clearer with colors (but I'm biased):

<img width="654" alt="Screen Shot 2022-05-02 at 7 07 30 PM" src="https://user-images.githubusercontent.com/1606434/166394956-ab01ab0f-0643-48fc-8384-2407ffb08dcd.png">

estebank commented:
Merging as per 

> Feel free to refine it if you want, or r=me if not.

@bors r=compiler-errors

bors commented:
:pushpin: Commit 4934a9e6430b3fcb98f77205f8a625c7ffebd628 has been approved by `compiler-errors`

<!-- @bors r=compiler-errors 4934a9e6430b3fcb98f77205f8a625c7ffebd628 -->
<!-- homu: {"type":"Approved","sha":"4934a9e6430b3fcb98f77205f8a625c7ffebd628","approver":"compiler-errors"} -->

bors commented:
:hourglass: Testing commit 4934a9e6430b3fcb98f77205f8a625c7ffebd628 with merge 8c727a798c153d9239d2378e139f9a5028efdc24...
<!-- homu: {"type":"BuildStarted","head_sha":"4934a9e6430b3fcb98f77205f8a625c7ffebd628","merge_sha":"8c727a798c153d9239d2378e139f9a5028efdc24"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6277548432?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6277548432?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6275961521?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/8c727a798c153d9239d2378e139f9a5028efdc24/checks/6275961521/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

compiler-errors commented:
@bors retry transient error on apple runner it seems

bors commented:
:hourglass: Testing commit 4934a9e6430b3fcb98f77205f8a625c7ffebd628 with merge 9add63257bd3968a7b7ea3372db9d548561e3396...
<!-- homu: {"type":"BuildStarted","head_sha":"4934a9e6430b3fcb98f77205f8a625c7ffebd628","merge_sha":"9add63257bd3968a7b7ea3372db9d548561e3396"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6284283425?check_suite_focus=true)
Approved by: compiler-errors
Pushing 9add63257bd3968a7b7ea3372db9d548561e3396 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6284283425?check_suite_focus=true"},"merge_sha":"9add63257bd3968a7b7ea3372db9d548561e3396"} -->

rust-timer commented:
Finished benchmarking commit (9add63257bd3968a7b7ea3372db9d548561e3396): [comparison url](https://perf.rust-lang.org/compare.html?start=1b2e0b60cc5aec81b48835bfc322c426e2f07cd1&end=9add63257bd3968a7b7ea3372db9d548561e3396).

**Summary**: 
- Primary benchmarks: 😿 relevant regressions found
- Secondary benchmarks: no relevant changes found

|            | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:----------:|:------------------------------:|:--------------------------------:|:-------------------------------:|:---------------------------------:|:------------------------:|
| count[^1]  | 2                              | 0                                | 0                               | 0                                 | 2                        |
| mean[^2]   | 0.5%                           | N/A                              | N/A                             | N/A                               | 0.5%                     |
| max        | 0.8%                           | N/A                              | N/A                             | N/A                               | 0.8%                     |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
